File: src/main/java/social/alone/server/user/UserEnrollService.java
package social.alone.server.user;

import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.util.Assert;
import social.alone.server.auth.FacebookUserInfoFetcher;
import social.alone.server.auth.oauth2.user.FacebookOAuth2UserInfo;

import org.springframework.transaction.annotation.Transactional;
import social.alone.server.infrastructure.S3Uploader;

import java.util.Optional;

@Service
@Transactional
@RequiredArgsConstructor
public class UserEnrollService {

    private final UserRepository userRepository;

    private final PasswordEncoder passwordEncoder;

    private final FacebookUserInfoFetcher facebookUserInfoFetcher;

    private final S3Uploader s3Uploader;

    public User enrollByEmailPassword(
            String email, String password, String name
    ) {
        Assert.notNull(email, "required");
        Assert.notNull(password, "required");
        Assert.notNull(name, "required");

        User user = User.builder()
                .name(name)
                .email(email)
                .password(passwordEncoder.encode(password))
                .build();

        return userRepository.save(user);
    }

    public User byFacebook(String accessToken){
        FacebookOAuth2UserInfo userInfo = facebookUserInfoFetcher.getUserInfo(accessToken);
//        userInfo.setS3ImageUrl(
//                s3Uploader.upload(userInfo.getId() + "_fb_profile.jpeg", userInfo.getImageUrl())
//        );
        return this.enrollByFacebook(userInfo);
    }

    private User enrollByFacebook(
            FacebookOAuth2UserInfo userInfo) {
        String email = userInfo.getEmail();
        Optional<User> byEmail = userRepository.findByEmailAndProvider(email, AuthProvider.facebook);
        return byEmail.orElseGet(() -> userRepository.save(new User(userInfo, AuthProvider.facebook)));
    }

}

File: src/main/kotlin/social/alone/server/user/service/UserEnrollService.kt
package social.alone.server.user.service

import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.Assert
import social.alone.server.auth.FacebookUserInfoFetcher
import social.alone.server.auth.anonymous.JoinByProfileRequest
import social.alone.server.auth.oauth2.user.FacebookOAuth2UserInfo
import social.alone.server.image.Image
import social.alone.server.image.ImageRepository
import social.alone.server.infrastructure.S3Uploader
import social.alone.server.user.domain.AuthProvider
import social.alone.server.user.domain.Profile
import social.alone.server.user.domain.User
import social.alone.server.user.repository.UserRepository
import java.util.*

@Service
@Transactional
class UserEnrollService(
        private val userRepository: UserRepository,
        private val passwordEncoder: PasswordEncoder,
        private val facebookUserInfoFetcher: FacebookUserInfoFetcher,
        private val imageRepository: ImageRepository
) {

    fun enrollByEmailPassword(
            email: String, password: String, name: String
    ): User {
        Assert.notNull(email, "required")
        Assert.notNull(password, "required")
        Assert.notNull(name, "required")

        val user = User(email, passwordEncoder.encode(password), Profile(name))

        return userRepository.save(user)
    }

    fun byFacebook(accessToken: String): User {
        val userInfo = facebookUserInfoFetcher.getUserInfo(accessToken)
        return this.enrollByFacebook(userInfo)
    }

    private fun enrollByFacebook(
            userInfo: FacebookOAuth2UserInfo): User {
        val email = userInfo.email
        val byEmail = userRepository.findByEmailAndProvider(email, AuthProvider.facebook)
        return byEmail.orElseGet { userRepository.save(User(userInfo, AuthProvider.facebook)) }
    }

    fun byProfile(request: JoinByProfileRequest): User {
        val fakeEmail = request.profile.name + "@anonymous.com"
        val byEmail = userRepository.findByEmailAndProvider(fakeEmail, AuthProvider.facebook)
        if (byEmail.isPresent) {
            return byEmail.get()
        }

        val profile = requestToProfile(request)
        val user = User(fakeEmail, null, profile)
        return userRepository.save(user)
    }


    private fun requestToProfile(request: JoinByProfileRequest): Profile {
        val image = request.profile.image?.id?.let { imageRepository.findById(it) } ?: Optional.empty()
        return Profile(request.profile.name, image.orElse(null))
    }

}
